### ISSUE TITLE
Yamux + secio data race

### ISSUE THREAD

**Issue by hsanjuan** [libp2p/go-libp2p#396]
I have found a yamux + secio data race when running tests and I'm a bit lost with this one:

```
==================
WARNING: DATA RACE
Write at 0x00c420d9eb5a by goroutine 310:
  gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux.header.encode()
      /usr/lib64/go/1.10/src/encoding/binary/binary.go:119 +0x1d9
  gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux.(*Stream).sendReset()
      /home/hector/go/src/gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux/stream.go:287 +0xe1
  gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux.(*Stream).Reset()
      /home/hector/go/src/gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux/stream.go:306 +0xac
  gx/ipfs/QmdjC8HtKZpEufBL1u7WxvQn78Lqq2Wk31NJS8WvFX3crB/go-libp2p-swarm.(*Stream).Reset()
      /home/hector/go/src/gx/ipfs/QmdjC8HtKZpEufBL1u7WxvQn78Lqq2Wk31NJS8WvFX3crB/go-libp2p-swarm/swarm_stream.go:111 +0x52
  gx/ipfs/QmUDzeFgYrRmHL2hUB6NZmqcBVQtUzETwmFRUc9onfSSHr/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler()
      /home/hector/go/src/gx/ipfs/QmUDzeFgYrRmHL2hUB6NZmqcBVQtUzETwmFRUc9onfSSHr/go-libp2p/p2p/host/basic/basic_host.go:228 +0x353
  gx/ipfs/QmUDzeFgYrRmHL2hUB6NZmqcBVQtUzETwmFRUc9onfSSHr/go-libp2p/p2p/host/basic.(*BasicHost).(gx/ipfs/QmUDzeFgYrRmHL2hUB6NZmqcBVQtUzETwmFRUc9onfSSHr/go-libp2p/p2p/host/basic.new
      /home/hector/go/src/gx/ipfs/QmUDzeFgYrRmHL2hUB6NZmqcBVQtUzETwmFRUc9onfSSHr/go-libp2p/p2p/host/basic/basic_host.go:154 +0x55
  gx/ipfs/QmdjC8HtKZpEufBL1u7WxvQn78Lqq2Wk31NJS8WvFX3crB/go-libp2p-swarm.(*Conn).start.func1.1()
      /home/hector/go/src/gx/ipfs/QmdjC8HtKZpEufBL1u7WxvQn78Lqq2Wk31NJS8WvFX3crB/go-libp2p-swarm/swarm_conn.go:113 +0xec

Previous read at 0x00c420d9eb5a by goroutine 131:
  crypto/cipher.fastXORBytes()
      /usr/lib64/go/1.10/src/crypto/cipher/xor.go:39 +0x131
  crypto/cipher.xorBytes()
      /usr/lib64/go/1.10/src/crypto/cipher/xor.go:60 +0x9e
  crypto/cipher.(*ctr).XORKeyStream()
      /usr/lib64/go/1.10/src/crypto/cipher/ctr.go:78 +0x2a1
  gx/ipfs/QmZ69yKFZ7V5hLCYFWzJF1Nz4dH1Es4LaE8jGXAFucLe3K/go-libp2p-secio.(*etmWriter).WriteMsg()
      /home/hector/go/src/gx/ipfs/QmZ69yKFZ7V5hLCYFWzJF1Nz4dH1Es4LaE8jGXAFucLe3K/go-libp2p-secio/rw.go:49 +0x1da
  gx/ipfs/QmZ69yKFZ7V5hLCYFWzJF1Nz4dH1Es4LaE8jGXAFucLe3K/go-libp2p-secio.(*etmWriter).Write()
      /home/hector/go/src/gx/ipfs/QmZ69yKFZ7V5hLCYFWzJF1Nz4dH1Es4LaE8jGXAFucLe3K/go-libp2p-secio/rw.go:34 +0x5a
  gx/ipfs/QmWBug6eBS7AxRdCDVuSY5CnSit7cS2XnPFYJWqWDumhCG/go-msgio.(*readWriter).Write()
      <autogenerated>:1 +0x82
  gx/ipfs/QmZ69yKFZ7V5hLCYFWzJF1Nz4dH1Es4LaE8jGXAFucLe3K/go-libp2p-secio.(*secureSession).Write()
      <autogenerated>:1 +0x7d
  gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux.(*Session).send()
      /home/hector/go/src/gx/ipfs/QmZt87ZHYGaZFBrtGPEqYjAC2yNAhgiRDBmu8KaCzHjx6h/yamux/session.go:398 +0xf3
```

**Comment by Stebalien** [comment]
This is a bug in yamux. It could have been a timed-out write or some kind of weirdness when shutting down a session. Both will cause yamux to abort a `waitForSendErr` call.

**Comment by Stebalien** [comment]
Interestingly, this *exact* issue is why I've been dragging my feet on correctly implementing write timeouts in mplex.

**Comment by Stebalien** [comment]
For context, the general issue here is that we *can't* cancel writes to things implementing `io.Write` (it doesn't take a context). Furthermore, canceling a write may leave some data written and some data not written.

However, yamux *must* support cancellation. It must be possible to reset a stream, it must be possible to set a write deadline on a stream, etc.

---

The solution here is, unfortunately, to copy. That is, in `Session.send`, we should copy the outgoing data into a temporary buffer (ideally packing a few "messages" into a single buffer) and *then* write it out to the connection.

The tricky part is that:

1. Once we *start* writing the data, we have to either *finish* writing the data or we have to kill the entire connection. Due to how yamux works, there's no way to "cancel" write after it has been started.
2. If the stream's write times out at a higher layer after we've started writing in the session, we *have* to tell the user that the write succeeded.

Basically, this all comes down to some *very* careful book keeping.

**Comment by Stebalien** [comment]
@raulk could you pick this up if/when you find time? Feel free to finish up anything you're currently working on first.

**Comment by raulk** [comment]
@Stebalien 👍

**Comment by raulk** [comment]
So the issue is indeed that each yamux stream keeps two `header`s at the stream level: the `sendHdr` and the `controlHdr`. They are each locked by a `sync.Mutex`, and two concurrent writes to the stream will be serialised by `waitForSendErr()`, so no issue on the surface.

This issue arises when `waitForSendErr()` returns with an error while the the underlying `conn.Write` (which in this case is wired to a SecIO transport) is still in progress. I _think_ what's happening here is a timeout. The alternative scenario is an error from `conn.Write()`, which would implicitly finish the transport write anyway.

So if a timeout is raised before the `conn.Write()` completes, the latter would still be underway while a new concurrent `Reset()` call is allowed to happen – which is what we're seeing here.

Since we make no copy of the `controlHdr` `header/byte[]` prior to calling `waitForSendErr()`, the second `Reset()` call tries to modify it and incurs in the data race.

Solutions I can think of:
1. Make a copy of the header in `send()` before engaging in transport IO (12 byte overhead). Not sure if we'll need to copy the body as well though, because there's a chance we can incur in the same issue if the application/protocol is reusing the underlying `Reader` across writes. We can add a note in the godoc stating that if the `Write` returns an `error`, the underlying `Reader` must be discarded. I'd need to assess the impact at the stream bookkeeping level, though. Not sure what a write that is seen by the other party, but failed in our end, would imply.
2. Terminate the connection as you said, which seems pretty drastic but may ultimately be necessary for 100% correctness.
3. Find a way to cancel the write.

**Comment by raulk** [comment]
@Stebalien forgot to mention you ^^

**Comment by Stebalien** [comment]
So, I think we have to do one of two things. Either:

1. Copy *everything*. This is the simplest and probably not *that* bad (we'd copy into a reusable buffer). The *real* issue here would be the memory overhead of *keeping* this buffer around (one per connection).
2. Set a deadline on the connection itself. However, even *if* we do this, we may write some bytes before we hit the deadline. If we do, we'll have to copy and continue writing the rest in the background. This is likely going to be extremely complicated. It may even be slower due to the additional syscalls required.

**Comment by raulk** [comment]
@Stebalien I think I've got a way forward here. I'm testing out a few things today. 

Just to clarify, regarding (2), is there a particular reason we should continue the write in the background? Currently an error from `conn.Write()` shuts down the connection: https://github.com/hashicorp/yamux/blob/master/session.go#L404 

In fact, I'd argue that most L7 apps (e.g. http clients) close the underlying connection (and maybe retry) on a write timeout. I don't think you have another way of recovering from this while preserving correctness, unless you introduce a backchannel to notify the app when a previously timed out session has resurrected.

But that's too far fetched. I'd suspect the moment you notify the app of a timeout, it's going to react and make internal state changes anyway.

**Comment by raulk** [comment]
Ok I think I might have misunderstood what conn.SetWriteDeadline does...

**Comment by Stebalien** [comment]
> Just to clarify, regarding (2), is there a particular reason we should continue the write in the background? Currently an error from conn.Write() shuts down the connection: https://github.com/hashicorp/yamux/blob/master/session.go#L404, and a WriteDeadline would take this circuit too.

A write deadline on the connection would but a write deadline on the stream shouldn't. FYI, we probably have write deadline issues all the way down (they're really painful to implement correctly, I'm actually *really* annoyed at golang for including them in the default connection interface).

> In fact, I'd argue that most L7 apps (e.g. http clients) close the underlying connection (and maybe retry) on a write timeout. I don't think you have another way of recovering while preserving correctness, unless you introduce a backchannel to notify the app when a previously timed out session has resurrected.


You're right, that's why we usually don't bother with them. However, we'd have to make sure to only kill the *stream*. The issue here is that we want to allow a user to set a deadline on a single stream so we need to finish the write (well, at least the write of the current "frame") before we can move on.

**Comment by raulk** [comment]
@Stebalien I agree that shutting down only the problematic stream is what's desirable. My concern is that more often than not, a connection error indicates a problem in the underlying transport which will likely affects all streams eventually. ill connection == ripple effect across streams. For sake or resilience and defensiveness, I much rather _contain_ this failure, than let it be.

**Proposal:** A transport failure should place the Session in **"grace period"** status with a failure counter. Upon the initial failure, we can be lenient, close the particular stream, and queue up a _prioritised packet_ to signal the stream `RST` for when the transport becomes healthy again, if at all. As streams continue to fail, we continue to do the former, until there comes a point where we simply kill the session altogether. Such failures could also gradually reduce the peer's score in the ConnMgr.

This would require deep changes to yamux, but I'm happy to moonlight on this ;-)

**Comment by Stebalien** [comment]
A connection *error*, yes. I'm just talking about stream timeouts. If the connection errors, we should just kill the connection (we already do that, don't we?). However, we should be able to set deadlines on individual streams and not kill the entire connection if we exceed the deadline on one stream.

**Comment by raulk** [comment]
Sorry @Stebalien, I'm probably a bit thick now (late here), but could you elaborate further on how can a stream time out by itself, without that timeout being caused by a write on the underlying connection?

I get that we want to specify individual deadlines on streams, but I'm trying to understand how that would interact with the existing [`ConnectionWriteTimeout`](https://github.com/hashicorp/yamux/blob/3520598351bb3500a49ae9563f5539666ae0a27c/session.go#L336), especially if the implementation relies on write deadlines on the shared connection.

Would the `ConnectionWriteTimeout` become a fallback value?

**Comment by raulk** [comment]
I see why it can be confusing. That _stream write deadline_ is used only when the stream's window is stalled, and we're waiting for a slide/increase. But if we're past that point, we've already called `conn.Write()`.

> Really, we should be killing the connection when the ConnectionWriteTimeout elapses. I'm not sure why it's not doing that.

`conn` is stored as a `io.ReadWriteCloser` which does not expose any deadline management facilities (but is handy for mock testing). So the plain calls to [`conn.Write()`](https://github.com/whyrusleeping/yamux/blob/master/session.go#L406) can stall forever.

In this instance, the [timeout set on an upper layer is firing](https://github.com/whyrusleeping/yamux/blob/master/session.go#L357) and reporting a timeout to the application, while the `conn.Write()` is still ongoing.

[Here's an initial stab at a fix.](https://github.com/raulk/yamux/commit/87496496be46373ae595acd4aa1730fa475945dc) As usual with socket testing, it's difficult to reproduce in a unit test. I'm currently stuck there.

That patch:
1. Uses `net.Conn.SetWriteDeadline()` to set timeouts when the underlying conn is a `net.Conn`, doing timer bookkeeping and handoff.
2. Falls back to previous logic when the transport is not a `net.Conn`, but improves it by adding yielding checkpoints in write loops (best-effort cancellation).

**Comment by Stebalien** [comment]
So, we're *killing* the connection on partial writes anyways, right? In that case, we can (I think) just kill the connection from `waitSendErr` if the connection times out. That should cancel any in progress writes anyways and simplify the patch significantly. I believe we only need fancy deadline dances if we want to somehow *avoid* killing the connection.

**Comment by raulk** [comment]
@Stebalien yeah, that makes sense. In the current state of affairs, the connection is not killed, but this seems like a bug. Errors from `conn.Write()` _do_ kill the connection, but because a deadline is not being set, the call can block forever. Setting a deadline there would return an error after the deadline, which would kill the connection. 

Indeed, having the timer in `waitSendErr` and killing the connection there would cancel the pending write anyway. It's simpler so I will go with that.

**Comment by raulk** [comment]
@Stebalien I set out to simplify the deadline handling, but then I remembered why it was necessary. 

1. Goroutines push writes via a buffered channel. So a timer expiring on `waitSendErr` could also mean that the channel buffer was full, or that the write was not consumed from the channel in a timely manner, hence IO never started for that write. Having many writes in that state close the connection at once would be unorderly.

2. We only want to kill the connection on partial writes, and in `waitSendErr` we don't have the number of bytes written, so we would need to make the connection participate in the dance by setting a deadline.

Opened a PR here: https://github.com/whyrusleeping/yamux/pull/10 with abundant comments to explain the solution.

**Comment by Stebalien** [comment]
@raulk has fixed this! :tada:

**Comment by hsanjuan** [comment]
Hello, I'm not sure this has been fixed. I very similar races popping up:

```
==================
WARNING: DATA RACE
Write at 0x00c0002d940a by goroutine 377:
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.header.encode()
      /usr/lib64/go/1.11/src/encoding/binary/binary.go:119 +0x1d9
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.(*Stream).sendReset()
      /home/hector/go/src/gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux/stream.go:288 +0x12e
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.(*Stream).Reset()
      /home/hector/go/src/gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux/stream.go:307 +0xac
  gx/ipfs/QmYyFNDLTSy6rW99K8vHPvFPLa5bB4zoy8gGZENoiaBN6R/go-libp2p-swarm.(*Stream).Reset()
      /home/hector/go/src/gx/ipfs/QmYyFNDLTSy6rW99K8vHPvFPLa5bB4zoy8gGZENoiaBN6R/go-libp2p-swarm/swarm_stream.go:116 +0x54
  gx/ipfs/QmSTaEYUgDe1r581hxyd2u9582Hgp3KX4wGwYbRqz2u9Qh/go-libp2p-net.FullClose()
      /home/hector/go/src/gx/ipfs/QmSTaEYUgDe1r581hxyd2u9582Hgp3KX4wGwYbRqz2u9Qh/go-libp2p-net/stream_util.go:25 +0x6f
  gx/ipfs/QmPZFUt4riSxc7PMLX3MGJ5o3uNufJmkTGXEx7SnUaFeP9/go-libp2p-floodsub.(*PubSub).handleSendingMessages()
      /home/hector/go/src/gx/ipfs/QmPZFUt4riSxc7PMLX3MGJ5o3uNufJmkTGXEx7SnUaFeP9/go-libp2p-floodsub/comm.go:78 +0x84f

Previous read at 0x00c0002d940a by goroutine 343:
  crypto/cipher.fastXORBytes()
      /usr/lib64/go/1.11/src/crypto/cipher/xor.go:39 +0x247
  crypto/cipher.xorBytes()
      /usr/lib64/go/1.11/src/crypto/cipher/xor.go:60 +0x9e
  crypto/cipher.(*ctr).XORKeyStream()
      /usr/lib64/go/1.11/src/crypto/cipher/ctr.go:86 +0x2af
  gx/ipfs/QmZUBPE2W68FpZgUSmhVyjVD55P5Bm51wjnU8hTH14jHpk/go-libp2p-secio.(*etmWriter).WriteMsg()
      /home/hector/go/src/gx/ipfs/QmZUBPE2W68FpZgUSmhVyjVD55P5Bm51wjnU8hTH14jHpk/go-libp2p-secio/rw.go:49 +0x1d9
  gx/ipfs/QmZUBPE2W68FpZgUSmhVyjVD55P5Bm51wjnU8hTH14jHpk/go-libp2p-secio.(*etmWriter).Write()
      /home/hector/go/src/gx/ipfs/QmZUBPE2W68FpZgUSmhVyjVD55P5Bm51wjnU8hTH14jHpk/go-libp2p-secio/rw.go:34 +0x5a
  gx/ipfs/QmcxL9MDzSU5Mj1GcWZD8CXkAFuJXjdbjotZ93o371bKSf/go-msgio.(*readWriter).Write()
      <autogenerated>:1 +0x8c
  gx/ipfs/QmZUBPE2W68FpZgUSmhVyjVD55P5Bm51wjnU8hTH14jHpk/go-libp2p-secio.(*secureSession).Write()
      <autogenerated>:1 +0x87
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.(*Session).send()
      /home/hector/go/src/gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux/session.go:450 +0x29a

Goroutine 377 (running) created at:
  gx/ipfs/QmPZFUt4riSxc7PMLX3MGJ5o3uNufJmkTGXEx7SnUaFeP9/go-libp2p-floodsub.(*PubSub).processLoop()
      /home/hector/go/src/gx/ipfs/QmPZFUt4riSxc7PMLX3MGJ5o3uNufJmkTGXEx7SnUaFeP9/go-libp2p-floodsub/pubsub.go:210 +0x13f0

Goroutine 343 (finished) created at:
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.newSession()
      /home/hector/go/src/gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux/session.go:123 +0x773
  gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux.Server()
      /home/hector/go/src/gx/ipfs/Qmdhe5BPbGaGutZ2E6LYGMm7FLNFeoFpgd5oLwE4hZBhWX/yamux/mux.go:79 +0xd9
  gx/ipfs/QmVwYCtShoL74Xi8TgEg9jXvHVVtWwZ3Hg1QZqQvQA9xji/go-smux-yamux.(*Transport).NewConn()
      /home/hector/go/src/gx/ipfs/QmVwYCtShoL74Xi8TgEg9jXvHVVtWwZ3Hg1QZqQvQA9xji/go-smux-yamux/yamux.go:66 +0x81
  gx/ipfs/QmRYdszNNq7ykPqavVNKMVyyjX59AcTisHqzussDfhwHkK/go-smux-multistream.(*Transport).NewConn()
      /home/hector/go/src/gx/ipfs/QmRYdszNNq7ykPqavVNKMVyyjX59AcTisHqzussDfhwHkK/go-smux-multistream/multistream.go:73 +0x1f4
  gx/ipfs/QmPM1NCCnMMY3GaPVkG2sTVh8Xys1gCPsk1X8Z1HuUJePs/go-libp2p-transport-upgrader.(*Upgrader).setupMuxer.func1()
      /home/hector/go/src/gx/ipfs/QmPM1NCCnMMY3GaPVkG2sTVh8Xys1gCPsk1X8Z1HuUJePs/go-libp2p-transport-upgrader/upgrader.go:118 +0xad
```

This is with `6.0.19`, which, I thought, included the fixes (correct me if I'm wrong).

**Comment by raulk** [comment]
Looking.

**Comment by raulk** [comment]
@hsanjuan what test is surfacing this?

**Comment by Stebalien** [comment]
It looks like we may need the `send` goroutine to signal that it has exited some how (e.g., close some channel, like `recv` dos).

**Comment by hsanjuan** [comment]
> @hsanjuan what test is surfacing this?

It happens in multiple cluster tests when doing -race . (https://ci.ipfs.team/blue/rest/organizations/jenkins/pipelines/IPFS/pipelines/ipfs-cluster/branches/PR-502/runs/6/nodes/11/steps/54/log/?start=0). The common thing is just that we create 5 libp2p hosts and do stuff with them and then shut them down.

**Comment by raulk** [comment]
I think what we're seeing now an unorderly close of streams/session. Protocols like pubsub are closing their streams, while at the same time the swarm is closing the sessions themselves. And yamux doesn't handle these concurrent closes very well.

I'm not convinced this issue _really_ affects production code, and fixing it would probably require rethinking much of yamux.

As an attempt to fix, I'll add some session shutdown checks in key spots.

**Comment by raulk** [comment]
Here's an attempt to bring some order in shutdowns: https://github.com/whyrusleeping/yamux/pull/22

@hsanjuan I am unable to reproduce these data races locally. Is there a way we can run your tests against this patch in Jenkisn without making a gx release?

**Comment by hsanjuan** [comment]
@raulk will attempt to run locally myself and report back. It it passes on my Raspberry Pi it's probably good. Thanks!

**Comment by raulk** [comment]
@hsanjuan: @Stebalien has added one more fix and we've made a new gx release upstream. Could you please update and re-test? Thanks!

**Comment by hsanjuan** [comment]
Thanks!! I will report back asap.

**Comment by hsanjuan** [comment]
Hey, I don't see yamux races anymore in our tests. Thank you!! I think this is fixed for the moment.
